Fix Python tests

jaen opened the pull request:
Makes `pytest .` ran from `nix develop` pass.

Fixing the `formatter` issue mentioned in https://github.com/nix-community/robotnix/issues/235 required updating the `repo` tool to the newest version. Since `mk_repo_file` uses a fork that adds additional commands to dump repo information, I have rebased it on top of the upstream v2.45 (see https://github.com/jaen/tools_repo/pull/1 for diff). Not sure where do we want to put this, though.

Atemu reviewed (commented):
The rest LGTM

jaen commented on scripts/mk_repo_file.py:
```diff
@@ -21,8 +21,8 @@
     "--quiet",
-    "--repo-url=https://github.com/danielfullmer/tools_repo",
-    "--repo-rev=9ecb9713ee5adba95120acbc0bfef1c77b02637f",
+    "--repo-url=https://github.com/jaen/tools_repo",
+    "--repo-rev=dca531f6d6e9fdcf00aa9d18f0153bd66a2e32ea",
```
That sounds reasonable,I'll try implementing it like this

jaen replied:
Turns out that doesn't help much, because the regular `gitRepo` derivation is basically just a launcher (c.f. https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/version-management/git-repo/default.nix#L32) that will then download the missing code from the interwebs (c.f. https://github.com/jaen/tools_repo/blob/main/repo#L669-L674). While it would be nice to make it reproducible, I'd say it's maybe out of scope for a simple test fix?

jaen replied:
Did overlay v2.45 in pkgs, though. Not sure if there's any downside to those versions not being the same, but might as well have them match.

jaen commented on flavors/lineageos/test_lineageos_updater.py:
```diff
@@ -3,27 +3,36 @@
 
```
I'd rather leave them in until a proper replacement is written (hopefully not in Python xD), personally. Unless you're very sure they are not testing anything useful?

Atemu replied:
I have touched the updater multiple times and not touched the test so...

jaen replied:
Nuked it,then

jaen commented on modules/envpackages.nix:
```diff
@@ -29,7 +29,15 @@ in
+      freetype # Needed by jdk9 prebuilt
+      fontconfig
+
+      # Goldfish doesn't need py2 anymore in Android 12+!
```
Okay not 100% sure how to trace this commit to Android version, but I think that's right

jaen commented:
FWIW I have successfully built the following ROM on this branch:
```
nix-build --arg configuration '{ device="lemonade"; flavor="lineageos"; androidVersion=13; apps.chromium.enable = false; webview.prebuilt.enable = false; webview.chromium.enable = false; webview.chromium.availableByDefault = false; }' -A factoryImg
```

Atemu commented:
Please note that the files you touched don't really influence build results very much but rather generating the sources.

jaen commented:
@Atemu I would think `envpackage.nix` and `release.nix` would have some influence on the build (having it potentially fail due to wrong Python version)? Which is why I figured building a ROM might make sense? But I don't know a lot about AOSP builds, so might've misunderstood something.

I have also just now modified the `repo` tool derivation to provide local sources, patched for repo2nix, like you asked for — it's kind of ugly, but at least doesn't require referencing my GitHub repo.

I ran update scripts for GrapheneOS (used `./update.sh SP2A.220405.004.2022041600`, files were identical) and for LineageOS (use `./update.sh lineage-20.0` files differed by timestamps and hashes, it also later failed because this repo does not exist: https://github.com/LineageOS/android_device_walmart_dopinder) and it seems to work.

jaen commented:
Re: above — added `flake-compat` to the flake and added a common entry-point for it, as I noticed later there were to places that used it (and I updated only one). It should be a bit more maintainable that way, hope you don't mind.

Atemu commented on pkgs/gitRepo/default.nix:
```diff
@@ -0,0 +1,70 @@
+    # Specify the patched checkout as the default version of repo
+    postFixup = ''
+      wrapProgram "$out/bin/repo" \
+        --set REPO_URL "file://$out/var/repo" \
+        --set REPO_REV "$(cat COMMITED_REPO_REV)" \
```
That's a great workaround, thanks!

Atemu commented on pkgs/gitRepo/default.nix:
```diff
@@ -0,0 +1,70 @@
+      mkdir -p var/repo
+      rsync -a $src/ var/repo/
+
+      (
+        export GIT_CONFIG_GLOBAL=$TMPDIR/.gitconfig
+        export GIT_CONFIG_NOSYSTEM=true
+
+        cd var/repo
+
+        git config --global --add safe.directory "$PWD"
+        git config --global user.email "nemo@nix"
+        git config --global user.name "Nemo Nix"
+
+        chmod +w -R .
+
+        git init
+        git add -A
+        git commit -m "Upstream sources"
+
+        git apply ${ ./00001-add-support-for-repo2nix.patch }
+        git add -A
+        git commit -m "Patch with repo2nix"
+
+        git log -n 1 --format="%H" > ../../COMMITED_REPO_REV
+      )
```
This is quite a hack. Would this perhaps be possible without producing a git repo?

jaen replied:
Unfortunately — as far as I can tell — `repo` needs it to be a git repo, as it basically clones itself out of it (see: https://gerrit.googlesource.com/git-repo/+/refs/heads/main/repo#674).

At first I tried to use `fetchFromGitHub` and `fetchFromGit` to produce a suitable git repo, but couldn't get it to work properly even with `leaveDotGit` – for example, patches didn't apply correctly (because they didn't get committed, I assume) and it _somehow_ poisoned the git branch used for manifest repo, though I didn't trace why exactly. And only then reached out for what you see, which worked.

For this PR I'd rather avoid patching `repo` further than it already was – if it needs a git repo to clone itself out of, I'd rather provide a git repo in the store at this stage, instead of figuring out how to patch it to not need git for self–bootstrap. It can always be improved in the future, if it turns out to be a problem.

There's possibly one more way to do this, as it seems that `repo` supports using a https://www.git-scm.com/docs/git-bundle to get its sources, but it doesn't really help with having to provide patched sources (either hosted, which I understand we didn't want, or massaged in a derivation, Like I've done). The only thing it would help with is that it would be a bundle in the store, not a git repo — which might be a bit cleaner conceptually, but not sure if solves any practical problem?

Atemu commented on pkgs/gitRepo/00001-add-support-for-repo2nix.patch:
Would be better to split this file into individual patches I think.

Atemu replied:
It'd be simpler to tell which patch was changed. You'd be importing/exporting as a series anyways, so there's no real overhead either.

jaen replied:
Sure, I'll amend the commit then.

Atemu commented on scripts/robotnix_common.py:
```diff
@@ -31,10 +31,10 @@ def save(filename: str, data: Any) -> None:
     """Get actual path to a Nix store path; supports handling local remotes"""
-    prefix = os.getenv('NIX_REMOTE');
+    prefix = os.getenv('NIX_REMOTE')
     if prefix and not prefix.startswith('/'):
         raise Exception('Must be run on a local Nix store.')
-    return f"{prefix}/{path}"
+    return os.path.join(*filter(None, [prefix, path]))
```
Sorry about that.

I'd prefer @eyJhb's patch from here: https://github.com/nix-community/robotnix/pull/228/files#diff-d2cdd059d29643c5fe1767c6dcf5203c0dc9b382d1322e3f0a4237eda6066246 though.

jaen replied:
Hmm, can you explain why? I've done it like this, so this function doesn't inadvertently turn a relative path to an absolute one rooted at `/`. Is doing that a desired behaviour of this function?

Atemu replied:
This function's purpose is to turn a nix store path to an absolute path that can be opened on the filesystem.

I've rewritten it, could you take a look?

jaen replied:
Seems reasonable; I'll re-test it locally just in case.

jaen replied:
Ok, after testing I'm kind of confused on what exactly `NIX_REMOTE` is. Nix itself seems to have that parameter (https://nix.dev/manual/nix/2.22/command-ref/env-common#env-NIX_REMOTE) but it doesn't seem to be that (because it's supposed to be a daemon socket).

From the context it looks like a mountpoint for some alternate store — but also not quite. When I added some logging to the code I get paths like `/nix/store/deadbeef-hudson`, and if you relativise them to root, you'll end up with paths like `/your/prefix/nix/store/deadbeef-hudson` which makes it not a store path IMO, I'd rather expect `/your/prefix/deadbeef-hudson` in this case, as if you specified `NIX_STORE_DIR` (https://nix.dev/manual/nix/2.22/command-ref/env-common#env-NIX_STORE_DIR) with that value.

But this is a bit of a documentation issue, I guess. I proposed a bit improved implementation in a separate commit — it leaves the `path` untouched if `NIX_REMOTE` is not set, otherwise it does the same logic but with `pathlib` which makes it a bit cleaner, as you don't have to worry about `.` not being resolved or knowing what is the root path to relativise the store path against.

Atemu replied:
`NIX_REMOTE` specifies a "remote" nix store. It's the equivalent of the `--store` flag.

If you set it to a local path, nix would then use the nix store under that prefix but it does not change the `NIX_STORE_DIR`. Build sandboxes are mounted such that the paths are available under `/nix/`.  
You can't directly run software off of such a store without a chroot but that's of little relevance for my robotnix builds.
The reason I use this is to store the 60GB+ build-time closure and hundreds of prefetched source dirs in a separate Nix store to my system's because I don't want to trash it. I explicitly do not want the `NIX_STORE_DIR` to change here.

(You can also point it at another machine via ssh in which case it builds on the remote machine but that doesn't work for the updater.) 

In the updater, we need to be able to access the `lineage.dependencies` file in a fetched repo and to do that while NIX_REMOTE points at some other location on the filesystem, you need this function. The returned string is not a store path but an absolute path that you can simply open and read.

Thanks for the pathlib patch, that looks a lot more robust. I don't have any experience with python to speak of, so this is much appreciated.

jaen replied:
Yeah, I did roughly understand what this function is for and it's a reasonable thing to want — I just wasn't sure what exactly it represents in terms of nix config/options/envvars (since nothing seemed to naively match it) and as such what I should expect the correct behaviour to be when testing it.

But it makes a lot more sense now that you explained what it's supposed to represent in terms of nix options/behaviour and I guess it actually indeed is `NIX_REMOTE` as described [here](https://nix.dev/manual/nix/2.22/command-ref/env-common#env-NIX_REMOTE), it's just that the description in the manual doesn't really make it clear that all values from [this list](https://nix.dev/manual/nix/2.22/command-ref/new-cli/nix3-help-stores.html) are valid and your explanation clarified that.

Atemu commented:
I'm currently cleaning this up a little and will force push your branch in a bit, just FYI.

jaen commented on pkgs/default.nix:
```diff
@@ -1,13 +1,7 @@
-  ) {
-    src =  ../.;
-  }).defaultNix.inputs,
+{ inputs ? (import ../flake/compat.nix).defaultNix.inputs,
```
So I don't forget later – seems I didn't add `system` here (or just pass empty `{}`, I guess):
```
  system ? builtins.currentSystem,
  inputs ? (import ../flake/compat.nix { inherit system; }).defaultNix.inputs,
  ...
```

jaen replied:
Fixed
